File: p3/p328/OddEvenList.java
package leetcode.p3.p328;

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class OddEvenList {
    public ListNode oddEvenList(ListNode head) {
        if (head == null) {
            return null;
        }
        if (head.next == null) {
            return head;
        }
        ListNode odd = head;
        ListNode even = head.next;
        ListNode temp = even;
        ListNode curr = head;
        while (curr.next != null && curr.next.next != null) {
            curr = curr.next.next;
            odd.next = curr;
            odd = odd.next;
            even.next = curr.next;
            even = even.next;
        }
        odd.next = temp;
        return head;
    }
}

File: p3/p328/odd_even_list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    """
    Dec 24, 2024 13:29
    """
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        odd_list = head
        even_list = head.next
        odd = head
        even = head.next
        while even.next and even.next.next:
            odd.next = even.next
            odd = odd.next
            even.next = even.next.next
            even = even.next

        if even.next:
            odd.next = even.next
            odd = odd.next
            even.next = None

        odd.next = even_list

        return odd_list


if __name__ == '__main__':
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))

    sol = Solution()

    print(sol.oddEvenList(head))
